rhobs/alerting/data_plane/prometheus.pipeline_alerts.yaml

```diff
@@ -183,3 +183,75 @@ spec:
+          for: 5m
+          labels:
+            severity: warning
+          annotations:
+            summary: Tekton Pipelines controller PipelineRun reconciler leases not distributed across more than one pod
+            description: >-
+              Tekton Pipelines controller PipelineRun reconciler leases not distributed across more than one pod
+              on cluster {{ $labels.source_cluster }}.
+            alert_team_handle: <!subteam^S04PYECHCCU>
+            team: pipelines
+            runbook_url: https://gitlab.cee.redhat.com/konflux/docs/sop/-/blob/main/o11y/alert-rule-PipelinesControllerLeasesNotDistributed.md
+        - alert: PipelinesControllerTaskRunRecLeasesNotDistributed
+          expr: (count by (source_cluster) (count by (lease_holder, source_cluster) (kube_lease_owner{lease=~"tekton-pipelines-controller.github.com.tektoncd.pipeline.pkg.reconciler.taskrun.reconciler..*", lease_holder!=""}))) <= 1
+          for: 5m
+          labels:
+            severity: warning
+          annotations:
+            summary: Tekton Pipelines controller TaskRun reconciler leases not distributed across more than one pod
+            description: >-
+              Tekton Pipelines controller TaskRun reconciler leases not distributed across more than one pod
+              on cluster {{ $labels.source_cluster }}.
+            alert_team_handle: <!subteam^S04PYECHCCU>
+            team: pipelines
+            runbook_url: https://gitlab.cee.redhat.com/konflux/docs/sop/-/blob/main/o11y/alert-rule-PipelinesControllerLeasesNotDistributed.md
+        - alert: PipelinesControllerResReqRecLeasesNotDistributed
+          expr: (count by (source_cluster) (count by (lease_holder, source_cluster) (kube_lease_owner{lease=~"tekton-pipelines-controller.github.com.tektoncd.pipeline.pkg.reconciler.resolutionrequest.reconciler..*", lease_holder!=""}))) <= 1
+          for: 5m
+          labels:
+            severity: warning
+          annotations:
+            summary: Tekton Pipelines controller ResolutionRequest reconciler leases not distributed across more than one pod
+            description: >-
+              Tekton Pipelines controller ResolutionRequest reconciler leases not distributed across more than one pod
+              on cluster {{ $labels.source_cluster }}.
+            alert_team_handle: <!subteam^S04PYECHCCU>
+            team: pipelines
+            runbook_url: https://gitlab.cee.redhat.com/konflux/docs/sop/-/blob/main/o11y/alert-rule-PipelinesControllerLeasesNotDistributed.md
+        - alert: PipelinesControllerPipelineRunRecLeasesHolderMissing
+          expr: count by (source_cluster) (kube_lease_owner{lease=~"tekton-pipelines-controller.github.com.tektoncd.pipeline.pkg.reconciler.pipelinerun.reconciler..*", lease_holder=""}) > 0
```

Comment:
With this expression, and because that metric will either give `1` or `_` (nothing), I wonder what will happen if you'd have a test case in which you had some `1`s and then had nothing. I'm not sure it won't alert.
i.e. have a test case in which you have a value of `1` for 6 minutes and then have `_` for about 10 minutes. When will the alarm stop firing?

Reply:
Added tests for that: One with `_x6 1x14` and another one with `1x6 _x14` - no alert in both cases.

Second case is obviously not a healthy state, but as I said above, I decided not try to detect that state by these rules. Hope it makes sense.

Reply:
> I decided not try to detect that state by these rules.
I understand what you mean now.

If that is the case, wouldn't be enough to have a simpler expressions such as this one:
```
kube_lease_owner{lease=~"tekton-pipelines-controller.github.com.tektoncd.pipeline.pkg.reconciler.pipelinerun.reconciler..*", lease_holder=""}
```

Reply:
Hmm, did not knew it is possible, but with that change, I'm getting so I guess I would need some aggregation anyway?
```
  FAILED:
    alertname: PipelinesControllerPipelineRunRecLeasesHolderMissing, time: 10m, 
        exp:[
            0:
              Labels:{alertname="PipelinesControllerPipelineRunRecLeasesHolderMissing", severity="warning", source_cluster="cluster01"}
              Annotations:{alert_team_handle="<!subteam^S04PYECHCCU>", description="Tekton Pipelines controller PipelineRun reconciler leases lease_holder is empty on cluster cluster01.", runbook_url="https://gitlab.cee.redhat.com/konflux/docs/sop/-/blob/main/o11y/alert-rule-PipelinesControllerLeasesNotDistributed.md", summary="Tekton Pipelines controller PipelineRun reconciler leases holder is missing", team="pipelines"}
            ], 
        got:[
            0:
              Labels:{alertname="PipelinesControllerPipelineRunRecLeasesHolderMissing", lease="tekton-pipelines-controller.github.com.tektoncd.pipeline.pkg.reconciler.pipelinerun.reconciler.00-of-04", namespace="openshift-pipelines", severity="warning", source_cluster="cluster01"}
              Annotations:{alert_team_handle="<!subteam^S04PYECHCCU>", description="Tekton Pipelines controller PipelineRun reconciler leases lease_holder is empty on cluster cluster01.", runbook_url="https://gitlab.cee.redhat.com/konflux/docs/sop/-/blob/main/o11y/alert-rule-PipelinesControllerLeasesNotDistributed.md", summary="Tekton Pipelines controller PipelineRun reconciler leases holder is missing", team="pipelines"},
            1:
              Labels:{alertname="PipelinesControllerPipelineRunRecLeasesHolderMissing", lease="tekton-pipelines-controller.github.com.tektoncd.pipeline.pkg.reconciler.pipelinerun.reconciler.01-of-04", namespace="openshift-pipelines", severity="warning", source_cluster="cluster01"}
              Annotations:{alert_team_handle="<!subteam^S04PYECHCCU>", description="Tekton Pipelines controller PipelineRun reconciler leases lease_holder is empty on cluster cluster01.", runbook_url="https://gitlab.cee.redhat.com/konflux/docs/sop/-/blob/main/o11y/alert-rule-PipelinesControllerLeasesNotDistributed.md", summary="Tekton Pipelines controller PipelineRun reconciler leases holder is missing", team="pipelines"},
            2:
              Labels:{alertname="PipelinesControllerPipelineRunRecLeasesHolderMissing", lease="tekton-pipelines-controller.github.com.tektoncd.pipeline.pkg.reconciler.pipelinerun.reconciler.02-of-04", namespace="openshift-pipelines", severity="warning", source_cluster="cluster01"}
              Annotations:{alert_team_handle="<!subteam^S04PYECHCCU>", description="Tekton Pipelines controller PipelineRun reconciler leases lease_holder is empty on cluster cluster01.", runbook_url="https://gitlab.cee.redhat.com/konflux/docs/sop/-/blob/main/o11y/alert-rule-PipelinesControllerLeasesNotDistributed.md", summary="Tekton Pipelines controller PipelineRun reconciler leases holder is missing", team="pipelines"},
            3:
              Labels:{alertname="PipelinesControllerPipelineRunRecLeasesHolderMissing", lease="tekton-pipelines-controller.github.com.tektoncd.pipeline.pkg.reconciler.pipelinerun.reconciler.03-of-04", namespace="openshift-pipelines", severity="warning", source_cluster="cluster01"}
              Annotations:{alert_team_handle="<!subteam^S04PYECHCCU>", description="Tekton Pipelines controller PipelineRun reconciler leases lease_holder is empty on cluster cluster01.", runbook_url="https://gitlab.cee.redhat.com/konflux/docs/sop/-/blob/main/o11y/alert-rule-PipelinesControllerLeasesNotDistributed.md", summary="Tekton Pipelines controller PipelineRun reconciler leases holder is missing", team="pipelines"}
            ]
```

Reply:
It depends on what you want to see. If you don't aggregate, it gives you more details on the lease that's missing a holder. Would that be valuable for whoever handles the alert?

If you prefer it aggregated then count by cluster, as you did, sounds good.

In any case, if you don't aggregate, I suppose you can have fewer data series for the test.